core/finality-grandpa/src/environment.rs

```diff
@@ -380,6 +380,19 @@ pub(crate) fn finalize_block<B, Block: BlockT<Hash=H256>, E, RA>(
 	// lock must be held through writing to DB to avoid race
 	let mut authority_set = authority_set.inner().write();
 
+	{
+		let (set_id, set_ref) = authority_set.current();
+		telemetry!(CONSENSUS_INFO; "afg.authority_set";
+			"hash" => ?hash,
+			"number" => ?number,
+			"authority_set_id" => ?set_id,
+			"authorities" => {
+				let authorities: Vec<String> =
+					set_ref.iter().map(|(id, _)| format!("{}", id)).collect();
+				format!("{:?}", authorities)
```

Comment:
The way authorities are collected here is not ideal and I'm not that happy with the solution. Happy to hear better suggestions!

Reply:
doing this for every single finalized block is kind of crazy. what's the purpose of this? Can't we just do it when the set has changed or on startup?

Reply:
The purpose is that we want to have the current authority set displayed in telemetry. If we send it only on startup+change this doesn't account for restarts of the telemetry server, which holds its state in memory and after a restart has no information about the authority set (and changes in the set which would retransmit this information are seldom).

I agree that the overhead of sending this with every finalized block is quite high. How about something like every era?

Do you have an idea how to compute the set less expensively?

Reply:
> If we send it only on startup+change this doesn't account for restarts of the telemetry server, which holds its state in memory and after a restart has no information about the authority set

If the server restarts then the client's connection should be broken. There should be a way to indicate that the most recent value of some key should be sent to telemetry when the connection is lost and re-established. Then we can do this only on start-up an change.

Reply:
Okay, I added commits which implement this. Could you please take another look?